src/SFA.DAS.EmployerFinance/Models/LevyDeclarationSaga.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using SFA.DAS.EmployerFinance.Messages.Events;
+
+namespace SFA.DAS.EmployerFinance.Models
+{
+    public class LevyDeclarationSaga : Entity
+    {
+        public int Id { get; private set; }
+        public LevyDeclarationSagaType Type { get; private set; }
+        public DateTime PayrollPeriod { get; private set; }
+        public long HighWaterMarkId { get; private set; }
+        public DateTime Created { get; private set; }
+        public DateTime? Updated { get; private set; }
+        public int ImportPayeSchemeLevyDeclarationsTasksCount { get; private set; }
+        public int ImportPayeSchemeLevyDeclarationsTasksCompleteCount { get; private set; }
+        public DateTime? ImportPayeSchemeLevyDeclarationsTasksFinished { get; private set; }
+        public int UpdateAccountTransactionBalancesTasksCount { get; private set; }
+        public int UpdateAccountTransactionBalancesTasksCompleteCount { get; private set; }
+        public DateTime? UpdateAccountTransactionBalancesTasksFinished { get; private set; }
+        public bool IsComplete { get; private set; }
+        public TimeSpan Timeout => TimeSpan.FromMinutes(2);
+
+        public LevyDeclarationSaga(DateTime payrollPeriod, IReadOnlyCollection<AccountPayeScheme> accountPayeSchemes)
+        {
+            Type = LevyDeclarationSagaType.All;
+            PayrollPeriod = payrollPeriod;
+            Created = DateTime.UtcNow;
+            HighWaterMarkId = accountPayeSchemes.OrderByDescending(aps => aps.Id).Select(aps => aps.Id).First();
+            ImportPayeSchemeLevyDeclarationsTasksCount = accountPayeSchemes.Select(aps => aps.EmployerReferenceNumber).Distinct().Count();
+            UpdateAccountTransactionBalancesTasksCount = accountPayeSchemes.Select(aps => aps.AccountId).Distinct().Count();
+            
+            Publish(() => new StartedProcessingLevyDeclarationsEvent(Id, PayrollPeriod, Created));
```

Comment:
What order does the EF SaveChanges and the NServiceBus Publish occur in their respective `UnitOfWork`'s, and is the order guaranteed?

If the Publish happens before the SaveChanges, this event (and also `StartedProcessingLevyDeclarationsAdHocEvent`) will be constructed before the `LevyDeclarationSaga` is saved, so the `sagaId` would always have a value of 0.

Reply:
Yeah `SaveChangesAsync` will be called on `EmployerFinanceDbContext` before the events are created and published 😄

The `IUnitOfWorkContext.AddEvent<T>(Func<T> messageFactory)` method was actually added for these kinds of scenarios i.e. to allow event creation to be deferred until a SQL Server identity has been determined.

Reply:
All already thought of 👍!